kernel/queue.c

```diff
@@ -102,29 +154,80 @@ void k_queue_insert(struct k_queue *queue, void *prev, void *data)
 	if (first_pending_thread) {
 		prepare_thread_to_run(first_pending_thread, data);
 		_reschedule(key);
-		return;
+		return 0;
 	}
 #endif /* !CONFIG_POLL */
 
-	sys_slist_insert(&queue->data_q, prev, data);
+	/* Only need to actually allocate if no threads are pending */
+	if (alloc) {
+		struct alloc_node *anode;
+
+		anode = z_thread_malloc(sizeof(*anode));
+		if (!anode) {
+			return -ENOMEM;
+		}
+		anode->data = data;
+		sys_sfnode_init(&anode->node, 0x1);
+		data = anode;
+	} else {
+		sys_sfnode_init(data, 0x0);
+	}
+	sys_sflist_insert(&queue->data_q, prev, data);
 
 #if defined(CONFIG_POLL)
 	handle_poll_events(queue, K_POLL_STATE_DATA_AVAILABLE);
 #endif /* CONFIG_POLL */
 
 	_reschedule(key);
+	return 0;
+}
+
+void k_queue_insert(struct k_queue *queue, void *prev, void *data)
+{
+	queue_insert(queue, prev, data, false);
```

Comment:
Yes because there's nothing I can do with this as the API itself is void.
It doesn't matter anyway since we pass alloc=false, and we only return -ENOMEM if an allocation is attempted.

Reply:
Yeah. We'll have to revisit these APIs returning void in the future. Failing silently and not giving an opportunity for code to handle that is not going to fly in functional safe scenarios